### ISSUE TITLE
Bug? Witness fails to fetch values from operations

### ISSUE THREAD

**Issue by soronpo** [milessabin/shapeless#738]
```scala
  def foo(r : Int)(implicit i : shapeless.Witness.Aux[r.type]) : Unit = {}
  val one = 1
  foo(one)
  foo(one + one) //fails
  val two = one + one
  foo(two)
```
Is this a bug/missing feature or is this not possible?

**Comment by milessabin** [comment]
This must be TLS with `-Yliteral-types`. Do you get the behaviour you expect with `ValueOf` instead of `Witness`?

**Comment by soronpo** [comment]
```scala
def foo(r : Int)(implicit i : ValueOf[r.type]) : Unit = {}
val one = 1
foo(one)
foo(one + one) //Error:(22, 6) No singleton value available for r.type.
```
I want to mention that I'm not expecting a constant value of 2, but I just need the tree of `one + one`.

**Comment by milessabin** [comment]
What do you mean by "I just need the tree of `one + one`"?

**Comment by soronpo** [comment]
I use `Witness` inside a macro to get the tree expression from the type. So all I require is the tree of `one+one`, if that is possible.

**Comment by soronpo** [comment]
I'll paste here my use case, as I explained on gitter.

```scala
def addExample(l : Int, r : Int)(implicit op : l.type + r.type): op.OutInt = op.value
val a = addExample(1, 2) //returns a : 3 = 3
val num = 2
val b = addExample(1,num) //returns b : Int = 3
```
My `+` operation is a def macro. It handles final values very well (singleton ops library). I expanded it so it will handle non-literal arguments to return a non-literal value. Everything works well, until I try to create a non-literal composition:
```scala
addExample(1,addExample(1,num) )  //fails
```
The reason this fails is what I gave in my simplified example. It is as if I were to write `addExample(1, num + num)`. I'm not that familiar with scalac internals. I want know if my macro can be modified in such a way so the type of `(num+num)` can be used to retrieve its tree. I currently use shapeless's `Witness` to retrieve the trees of non-literals and it fails as I demonstrated in the simplified example.
